### PR TITLE
Provide benchmarking tools for MOOSE

### PR DESCRIPTION
**Pull request by rwcarlsen**
This PR includes tools for benchmarking MOOSE in a consistent and trackable way.  We will be hooking this up with CIVET to run on the master branch and provide performance trends on mooseframework.org.  Mostly addresses #9265.

The new benchmarks directory holds files/scripts relevant to benchmarking and tracking MOOSE performance.  The
primary tool provided is the ``benchmark.py`` script which is used to run benchmarks, generate
trend visualizations, and look at stats comparing bencharks between various revisions.  To run all
benchmarks, just do:

```
$ ./benchmark.py --run
```

The tool reads a ``bench.list`` text file that specifies which input files should be run and
corresponding (benchmark) names for them along with any optional arguments.  The ``bench.list``
file has the following format:

```
[benchmark-name]|[binary]|[input-file]|[optional-args]
...
```

Note that the ``|`` characters are literal bars used as field delimiters.  Each line defines a
named benchmark by specifying paths to a binary and input file (paths rooted at the base of the
MOOSE repository) along with optional arguments that are appended to the end of the invocation of
the binary.  A sample ``bench.list`` file might look like this:

```
simple diffusion (refine3)|test/moose_test-opt|test/tests/kernels/simple_diffusion/simple_diffusion.i|Mesh/uniform_refine=3
simple diffusion (refine4)|test/moose_test-opt|test/tests/kernels/simple_diffusion/simple_diffusion.i|Mesh/uniform_refine=4
simple diffusion (refine5)|test/moose_test-opt|test/tests/kernels/simple_diffusion/simple_diffusion.i|Mesh/uniform_refine=5
...
```

When benchmarks are run, the binaries specified in ``bench.list`` must already exist.  Benchmark
data are then stored in a sqlite database (default name ``benchmarks.sqlite``).  You can specify
the minimum number of runs for each benchmark problem/simulation with the ``--min-runs`` (default
10).  Each benchmark will be run as many times as possible within 1 minute (customizable via the
``--cum-dur`` flag) or the specified minimum number of times (whichever is larger). 

MOOSE revisions with available benchmark data can be listed (from the database) by running:

```
$ ./benchmark.py --list-revs
44d2f3434b3346dc14fc9e86aa99ec433c1bbf10	2016-09-07 19:36:16
86ced0d0c959c9bdc59497f0bc9324c5cdcd7e8f	2016-09-08 09:29:17
447b455f1e2d8eda649468ed03ef792504d4b467	2016-09-08 09:43:56
...
```

To look at stats comparing benchmark data from two revisions, run:

```
$ ./benchmark.py # defaults to using the most recent two revisions of benchmark data
-------------------------------- 871c98630c98 to 38bb6f5ebe5f --------------------------------
          benchmark               old (sec/run)     new (sec/run)    speedup (pvalue,nsamples)
----------------------------------------------------------------------------------------------
    simple diffusion (refine3):      0.408034          0.408034          ~   (p=0.996 n=36+36)
    simple diffusion (refine4):      1.554724          1.561682          ~   (p=0.571 n=10+10)
    simple diffusion (refine5):      6.592326          6.592326          ~   (p=0.882 n=4+4)
----------------------------------------------------------------------------------------------

$ ./benchmark.py -old 44d2f34 -new 447b455 # or specify revisions to compare manually
------------------------------------- 44d2f34 to 447b455 -------------------------------------
          benchmark               old (sec/run)     new (sec/run)    speedup (pvalue,nsamples)
----------------------------------------------------------------------------------------------
    simple diffusion (refine3):      0.416574          0.411435        -1.2% (p=0.000 n=37+37)
    simple diffusion (refine4):      1.554724          1.497379        -3.7% (p=0.000 n=10+11)
    simple diffusion (refine5):      6.553244          6.360004        -2.9% (p=0.030 n=4+4)
----------------------------------------------------------------------------------------------
```

To generate visualizations, run:

```
$ ./benchmark.py --trends
```

This will generate an svg box plot for each benchmark over time/revision in a ``trends``
subdirectory.  An ``index.html`` file is also generated that embeds all the svg plots for
convenient viewing all together in a browser.

### CODE REVIEW

**Comment by friedmud** [comment]
This is looking good.  My only real issue with it is the `bench.list` file... for a few reasons:

1.  It's not like anything else in MOOSE.  We don't use that sort of input file syntax anywhere else.  I think it should be in GetPot format just like everything else (if it should exist at all).
2.  I don't like the specification of the binary.  The binary should get inferred from where you are currently running and the `METHOD` environment variable just like everything else.
3.  If we need a `bench.list` then there should be one for each application / module so that more that just MOOSE can use this (like all applications for instance).
4.  I still don't think `bench.list` is necessary - we should be able to encode within the normal `tests` files which tests should be used for benchmarking and what those benchmarking parameters are.  Having to have a separate file listing tests creates a maintenance burden.  When you add/modify/remove tests you have to remember to go edit `bench.list`...

In short: `run_benchmarks` should work VERY similarly to `run_tests`.  In fact: I still believe that it should be run like `./run_tests --benchmark --blah --di --bleh`... which would make for seamless roll out to all applications.

**Comment by YaqiWang** [comment]
This is a great move. We want to have something similar in applications. Also can we also track the memory consumption?

**Comment by rwcarlsen** [comment]
Point 1: I can switch to getpot - that is fine.

Point 2: But for benchmarking, we don't ever want to run with a non-optimized binary.  So just going based on METHOD would be dangerous.  Also, specifying the binary explicitly makes it really easy to provide the functionality you want in your point 3.

Point 3: I think it's nice to just be able to run all the benchmarks without having to go to a bunch of different directories one-by-one (e.g. one for each module, moose_test, etc.).  These things will be run by CIVET all together and it will be nice to be able to keep track of and maintain them together as a group.  For example - if the benchmarks are taking too long and we need to remove/shorten some, I can go to one place and just adjust several rather than having to hunt around and edit 10 files.  Also, as currently implemented, this does work for modules and other apps (except I need to push up a small fix for path-handling).

Point 4: I disagree - I think scattering around marks indicating which tests should be benchmarked creates more maintenance burden.  Also, users often copy the tests files and we don't want things inadvertently becoming benchmarks.

**Comment by andrsd** [comment]
Why not to have something like `type = Benchmark` in `tests` file?  I do not think a flag would be good, because (1) copy and paste problems, (2) does not make sense for things like Jacobian testers.

@rwcarlsen Why do you think having benchmarks in `tests` would have higher maintenance? Most diffs there are adding new tests, not really changing the existing ones.  Also, do you expect benchmarks to fail like tests? Because if a test fails, we need to do something. With benchmarks, we just run and measure, right? Or do we plan something more?

**Comment by friedmud** [comment]
`type = benchmark` in the `tests` files seems like a decent option.

That would let you fully specify exactly what to do to benchmark that test...

**Comment by friedmud** [comment]
Another good thing about `type = benchmark` is that it would be faster: we wouldn't need to do all of the things we currently do for a test (like running exodiff).

**Comment by permcody** [comment]
@rwcarlsen - I know you still want to build stand alone utilities because you believe that small compartmentalized utilities are easier to maintain. While this thinking isn't misguided, our experience has shown us that it's possible to create an ecosystem of tools that have common roots. More importantly than all of that, user experience and cohesiveness is key. The TestHarness was designed to look and feel just like MOOSE. We have python InputParameters, a Factory, a Warehouse, GetPot input files, etc., and our users love that. There is real value in a common ecosystem of tools.

Looking at the argument evolving here, I tend to stand with Derek on most things. I like David's suggestion of creating a new Tester type which would all but eliminate copy'n'paste problems. Yes, I know the TestHarness seems like a big piece of code but it actually already does a lot of things right that you may end up eventually re-inventing over time as you find you need them with the benchmarking tool: The TestHarness finds "input" files, tests system configurations, handles killing long running and broken tests or executables. You will need to recode each of these things and possibly more over time with the maintenance of this new stand-alone tool. It builds the right command line for parallel, and threaded runs, it does so many things already. In fact our existing test timing tool, already makes use of the TestHarness and the timing/database portion of that works just fine. It's the quality of the testing and the fact that we don't repeat tests that really causes issues. If you take a look at the TestHarness, you'll see that it does all this in a fraction of the code you just wrote.

I think we'd get a bigger bang for the buck if we pull our resources together to build a common tool. Besides your experience with software would help us smooth out the rough edges with some of the existing problems (like the parser). I'd value your input into the redesign of the TestHarness/Scheduler that Jason is working on. I think it's really cool and could easily be extended (with less code than you've already written) to perform the timing function.

**Comment by permcody** [comment]
We don't even have to put `benchmark` in the `tests` file. The TestHarness is designed to read different spec files so we could create separate files if we don't want users polluting the existing test specs. I think I would prefer this. We already use this capability extensively for the unit tests. Take a look at all the extension-less files here:
https://github.com/idaholab/moose/tree/devel/test/tests/test_harness

This means we can either scatter files, or not. It's already been coded and tested for us.

**Comment by rwcarlsen** [comment]
@friedmud - You have some good points, and I also have some - and I still disagree.

**Comment by andrsd** [comment]
> We don't even have to put benchmark in the tests file. The TestHarness is designed to read different spec files so we could create separate files if we don't want users polluting the existing test specs

We could go even further and put benchmarks into `benchmarks` dir with some structure underneath.

**Comment by friedmud** [comment]
Putting benchmarks in a `benchmarks` file right next to the normal `tests` file in a test directory makes sense to me.  It's still kept with the files it's referencing: but is nicely separated.

Reusing all of the execution structure of the current test harness makes sense.  We've grown all of that over many years... and we all know that getting it right is tricky.

**Comment by permcody** [comment]
Just for the record, here's the existing TestHarness based timer in it's entirety. It has been working for several years with zero maintenance because it's part of the testing infrastructure:
https://github.com/idaholab/moose/blob/devel/python/TestHarness/TestHarness.py#L992-L1064

The TestHarness is getting cleaned up with @milljm's current work so that it's function is more clearly defined. It'll still handle the finding of "input" specs and setting up the new pluggable "Scheduler" system. We are starting off with a new threaded Scheduler to parallelize not only the running, but also the post processing step in the TestHarness. This speeds up the TestHarness even more. The Testers are also pluggable. I would envision a benchmarking system as being a new Scheduler and a new Tester.

The Scheduler would be a simple serial running. Get a single job, run it and wait to keep the system quiet. The Tester would run a job several times to get the timings you need. All this would be stored as output in the Tester object, packaged up and sent back down through the Scheduler to the TestHarness where you would insert those values in the database. Despite the lengthy description, with the base objects doing 90% of the work, this would be very simple tasks to perform. 

Let's get the interface of this PR in the state we want it in, then we can merge it in and start looking at the data (So let's make sure the database structure is what we want and is extensible to store timings from different apps, maybe even different modes (is there any utility in doing this)? 

Now that you are using GetPot, it should be easier to transition to a TestHarness based system if we deem that worthwhile.

File: bench.list [hunk 1]
```diff
@@ -0,0 +1,23 @@
+[benchmarks]
+    [./simple_diffusion_refine3]
+        binary = test/moose_test-opt
```
Comment (permcody) [inline bench.list]: I'm OK with this right now since it's fairly stable but I'd prefer that we just pick it up. Let's not change this if we go with a version of this current script because you'll just be recoding logic that already exists in the TestHarness. If we switch systems, this will become automatic.

File: bench.list [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+        infile = test/tests/kernels/simple_diffusion/simple_diffusion.i
+        args = 'Mesh/uniform_refine=3'
```
Comment (permcody) [inline bench.list]: In the TestHarness input files we use `cli_args`. Consistency

File: benchmarks/benchmark-history.sh [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+4f4ce86
+ed7e7fc
+bc4e4bd
+e5801ef
+f2d41b2
+345c7a5
+a122953
+447b455
+86ced0d
+44d2f34'
```
Comment (permcody) [inline benchmarks/benchmark-history.sh]: This is just a bootstrap script right? It becomes automated under CIVET?

File: scripts/benchmark.py [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+        self.perflog = []
+
+    def run(self):
+        cmdpath = os.path.abspath(os.path.join(self.rootdir, self.executable))
+        infilepath = os.path.abspath(os.path.join(self.rootdir, self.infile))
+        cmd = [cmdpath, '-i', infilepath]
+        if self.args is not None:
+            cmd.extend(self.args)
+        cmd.append('Outputs/console=false')
+        cmd.append('UserObjects/perflog/type=PerfLogDumper')
```
Comment (permcody) [inline scripts/benchmark.py]: Might want to pull these "magic" options out and make them more prominent for maintenance.

File: scripts/benchmark.py [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+        self.infile = infile
+        self.args = args
+        self.dur_secs = 0
+        self.user_dur_secs = 0
+        self.perflog = []
+
+    def run(self):
+        cmdpath = os.path.abspath(os.path.join(self.rootdir, self.executable))
+        infilepath = os.path.abspath(os.path.join(self.rootdir, self.infile))
+        cmd = [cmdpath, '-i', infilepath]
```
Comment (permcody) [inline scripts/benchmark.py]: This is rather limited. The Tester base class handles building the right command up and can deal with a variety of MPI/threading environment. I'm just pointing that out now because it's already done.

File: scripts/benchmark.py [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+        cmd.append('UserObjects/perflog/type=PerfLogDumper')
+
+        tmpdir = 'tmp'
+        shutil.rmtree(tmpdir, ignore_errors=True)
+        os.makedirs(tmpdir)
+
+        with open(os.devnull, 'w') as devnull:
+            p = subprocess.Popen(cmd, cwd=tmpdir, stdout=devnull)
+            start = time.time()
+            (_, retcode, ru) = os.wait4(p.pid, 0) # do this hack to retrieve the usertime of the subprocess
```
Comment (permcody) [inline scripts/benchmark.py]: What if cmd never terminates? It happens, it happens a lot more than we'd like to admit.

File: scripts/benchmark.py [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+    return iqra
+
+
+class BenchmarkDB:
+    def __init__(self, fname):
+        CREATE_BENCH_TABLE = '''CREATE TABLE IF NOT EXISTS benchmarks
+        (
+          id INTEGER PRIMARY KEY AUTOINCREMENT,
+          name TEXT,
+          executable TEXT,
```
Comment (permcody) [inline scripts/benchmark.py]: I don't think this is the right schema. Does it save the full path or just the name? We might want a single app with methods. I don't know that's the right way to do it, but unless you are stripping it all down this might not be exactly what you want. When you start running with MPI, it'll have extra junk in front of the executable.

**Review by permcody — COMMENTED** [review]
This is getting better. I made a few comments, and hopefully you'll be amenable to making this work with the TestHarness. In fact, you could help me review Jason's PR to create this whole scheduler concept.

File: scripts/benchmark.py [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+    def run(self):
+        cmdpath = os.path.abspath(os.path.join(self.rootdir, self.executable))
+        infilepath = os.path.abspath(os.path.join(self.rootdir, self.infile))
+        cmd = [cmdpath, '-i', infilepath]
+        if self.args is not None:
+            cmd.extend(self.args)
+        cmd.append('Outputs/console=false')
+        cmd.append('UserObjects/perflog/type=PerfLogDumper')
+
+        tmpdir = 'tmp'
```
Comment (brianmoose) [inline scripts/benchmark.py]: Is there a reason you are using this tmp directory and not an auto generated one?
Especially since `tmp` is a rather common name and you are blowing it away....

File: scripts/benchmark.py [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+        infile = child.params['infile'].strip()
+        args = ''
+        if 'args' in child.params:
+            args = child.params['args']
+        benches.append(BenchEntry(benchname, ex, infile, args))
+    return benches
+
+if __name__ == '__main__':
+    find_moose_python()
+    from FactorySystem.ParseGetPot import readInputFile
```
Comment (brianmoose) [inline scripts/benchmark.py]: Just curious, why is this here and not with the other imports? Or in the function where `readInputFile` is used?

**Review by brianmoose — CHANGES_REQUESTED** [review]

File: scripts/benchmark.py (hunk 8 above)
Comment (permcody) [inline scripts/benchmark.py]: We went through several variations and figured out several things that don't work reliably. The existing code in RunParallel is rock solid.

File: benchmarks/benchmark-history.sh [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+
+for commit in $commits; do
+    git checkout $commit \
+        && (cd ../test && make -j60) \
+        && ./benchmarkk.py --run --benchlist bench.listt
```
Comment (rwcarlsen) [inline benchmarks/benchmark-history.sh]: This script will probably be removed before merge - it was just for testing/debugging.  The stutter was intentional to use a non-version controled version of the file that stayed around when checking out older commits in MOOSE that didn't have the actual bench.list.

File: scripts/benchmark.py [hunk 11]
```diff
@@ -0,0 +1,475 @@
...
+import shutil
+import collections
+import csv
+import matplotlib.pyplot as plt
+import jinja2
+import pprint
+
+def find_moose_python():
+    moosedir = os.path.abspath(os.path.join(os.path.dirname(__file__), '..'))
+    if os.environ.has_key('MOOSE_DIR'):
```
Comment (rwcarlsen) [inline scripts/benchmark.py]: Copy-pasted code.  Fixed.

File: scripts/benchmark.py (hunk 9 above)
Comment (rwcarlsen) [inline scripts/benchmark.py]: Because we need to add MOOSE_DIR/python to the list of dirs python hunts for modules in since this is a script and doesn't live in the python directory.

File: scripts/README.md [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+```
+$ ./benchmark.py --run
+```
+
+The tool reads a ``bench.list`` text file that specifies which input files should be run and
+corresponding (benchmark) names for them along with any optional arguments.  The ``bench.list``
+file has the following format:
+
+```
+[benchmark-name]|[binary]|[input-file]|[optional-args]
```
Comment (rwcarlsen) [inline scripts/README.md]: done.

File: scripts/benchmark.py (hunk 7 above)
Comment (rwcarlsen) [inline scripts/benchmark.py]: It's just the moose app binary name (plus path rooted from the bench.list dir)

File: bench.list (hunk 2 above)
Comment (rwcarlsen) [inline bench.list]: fixed.

File: benchmarks/benchmark-history.sh (hunk 3 above)
Comment (rwcarlsen) [inline benchmarks/benchmark-history.sh]: Correct.

File: scripts/benchmark.py (hunk 8 above)
Comment (rwcarlsen) [inline scripts/benchmark.py]: Fixed to use ``tempfile``.

File: scripts/benchmark.py (hunk 4 above)
Comment (rwcarlsen) [inline scripts/benchmark.py]: done

File: scripts/benchmark.py (hunk 6 above)
Comment (rwcarlsen) [inline scripts/benchmark.py]: timeout added.

File: bench.list (hunk 2 above)
Comment (permcody) [inline bench.list]: Fixed where? I guess you said this file was getting deleted so it's not fixed?

File: scripts/benchmark.py [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+        raise ValueError('benchmark list file "{}" does not exist'.format(benchlist))
+
+    benches = []
+    root = readInputFile(benchlist).children['benchmarks']
+    for benchname in root.children_list:
+        child = root.children[benchname]
+        ex = child.params['binary'].strip()
+        infile = child.params['infile'].strip()
+        args = ''
+        if 'cli_args' in child.params:
```
Comment (permcody) [inline scripts/benchmark.py]: OK

**Comment by rwcarlsen** [comment]
What are next steps for this?

File: scripts/benchmark.py (hunk 7 above)
Comment (permcody) [inline scripts/benchmark.py]: I'd prefer to throw away the path, I don't think it adds value and could potentially (although unlikely) add issues. However, I do think it's worth splitting off the suffix from this before storing it in the database. Reason: I think it may be useful (or somebody will eventually ask us for it) to compare performance of different methods. We might use this to justify moving from "opt" as the default to "oprof" for workstations for instance, or maybe we can convince developers that "devel" is really where it's at for developing because it's not really slower. Having it as a separate field would make querying the same application for different methods much easier.

File: scripts/benchmark.py [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+
+class BenchmarkDB:
+    def __init__(self, fname):
+        CREATE_BENCH_TABLE = '''CREATE TABLE IF NOT EXISTS benchmarks
+        (
+          id INTEGER PRIMARY KEY AUTOINCREMENT,
+          name TEXT,
+          executable TEXT,
+          input_file TEXT,
+          timestamp INTEGER,
```
Comment (permcody) [inline scripts/benchmark.py]: Databases can usually fill this field in automatically. I guess if you really don't want to, that's fine, just pointing that out.

**Review by permcody — CHANGES_REQUESTED** [review]
I left one more comment on the database schema, because I think it would be useful to store the application and method separately so that we could potentially compare different methods on the same app.

However, in the bigger picture, I still want to integrate this tool into the TestHarness. The existing TestHarness already does what you've reimplemented here (i.e. it has the ability to scan through directories, find files, execute the binary with cli_args and store those results in a DB). If you are antsy to get started with timing, I'm all for that, since the TestHarness may be in development for a little while longer. Can you at least agree to build your own scheduler once the new system is online? I don't see any real issues with taking this code and extending what we have. It'll unify the interface just that much more and make everyone happy.

As for now, let's store the method separately so we start with the right schema.  Thanks

File: scripts/benchmark.py (hunk 7 above)
Comment (rwcarlsen) [inline scripts/benchmark.py]: I added separate fields for the binary name and method.  I think the path is useful - it just keeps more information - which is easy to ignore if we don't want it.

**Comment by rwcarlsen** [comment]
I'd be happy to work on better integration with the test harness when the new design gets settled.  I also added separate fields for binary name and method to the db schema.

**Comment by permcody** [comment]
@rwcarlsen - I'd really like to see this tool as a new TestHarness `Scheduler`. It should be very simple to create a `RunSerial` class with you plotting capability that's proposed here. The upshot being that we can reuse the tree traversal, tester creation, and reporting capability but use your scheduler to run each test serially and several times.

**Comment by permcody** [comment]
@rwcarlsen - after you put out fires caused by the new parser. Let's revisit this and get it merged in. We need to poise ourselves for being able to work on optimization very soon.

**Comment by permcody** [comment]
Reworking in the TestHarness, Closing this PR.
